if `arrow_shift = :end`, move arrows to surface of destination node

hdavid16 opened the pull request:
This PR changes the behavior when `arrow_shift = :end`. Instead of having the center of the arrow marker on the center of the destination node, the code now calculates the position along the edge such that the tip of the arrow marker will touch the tip of the destination node marker. Note: currently works properly for `Circle` node marker and `Arrow` arrow marker since these are of base size `1x1`. 

`radius_px` can be extended for other marker types that do not have a `1x1` base size.

close #107 
close #101 
close #109 
close #112 
and update docs

To do:
- [x] test with interactive plot (i.e., drag node)

hdavid16 commented:
Changed the default node and arrow markers so that their size matches their pixels (see https://docs.makie.org/v0.19/examples/plotting_functions/scatter/index.html#markersize)

hdavid16 commented:
To do: add example in docs

hexaeder commented on src/recipes.jl:
```diff
@@ -196,6 +200,10 @@ function Makie.plot!(gp::GraphPlot)
     end
+    # function which projects the pixels to a point in data space
+    to_point = @lift (px) -> begin
+        px ./ ($to_px(Point(1,1)) - $to_px(Point(0,0)))
```
I think the `to_point` name and comment is misleading: if I am not mistaken this only contains the "scaling" but not the offset part of the linear transformation between px and dataspace. What about rotation? But might be fine to just change the name, i think it is not needed to have an actual function to go from px to dataspace.

hdavid16 replied:
Thanks for bringing this up. I don't know how transformations work, so I'm not sure how to address rotation. I will change the name for now if that works...

hdavid16 replied:
Changed to `scale_px`

hexaeder commented on src/recipes.jl:
```diff
@@ -213,21 +221,22 @@ function Makie.plot!(gp::GraphPlot)
-    arrow_pos = @lift if !isempty($edge_paths)
-        broadcast(interpolate, $edge_paths, $(gp.arrow_shift))
+    # plot arrow heads
+    arrow_shift = @lift update_arrow_shift!(graph[], gp, $edge_paths, $(gp.arrow_shift), $to_angle, $to_point)
```
I think the function is not mutating so shouldn't have a `!`. A mutating version with observables would look something like that:
```julia
arrow_shift = Observable(zeros(Float32, ne(gp))
onany(edge_paths, gp.arrow_shift, topoint) do ep, as, tp
    update_arrow_shift!(arrow_shift, graph[], ep, as, toangle[], tp)
end
```
(in that case you'd need to resize the `arrow_shift` to `ne(graph)` within the function)
Also you don't need to add an explicit observable-dependency on `toangle`, because if `toangle` updates it will trigger `topoint` and otherwise you'll trigger this function twice.

hdavid16 replied:
thanks. fixed

hdavid16 replied:
I tried removing the `$` from `to_angle`, but get the error:
ERROR: MethodError: objects of type Observable{GraphMakie.var"#1906#1926"{GraphMakie.var"#1904#1924"{Scene}}} are not callable

This happens because I call the function to_angle inside the `update_arrow_shift` function. How do we deal with this so that we don't trigger the function twice?

hdavid16 replied:
Ahh, never mind. I see the bracket next to it so that the function is passed and not the observable. Fixed it now.

hexaeder commented on src/recipes.jl:
```diff
@@ -654,3 +663,38 @@ function update_discretized!(disc, pathes)
+    arrow_shift = Vector{Float32}(undef, ne(g))
+
+    #node attr
+    node_marker = gp.node_marker[] 
```
`node_marker`, `arrow_marker` and `arrow_size` might be vectors, so you need to lookup using `getattr` for each destination node inside the loop.

hdavid16 replied:
ok, fixed

hexaeder reviewed (commented):
I am not totally sure about the choice `arrow_shift=:end` vs `arrow_shift=1`. I tend to prefere `:end` because otherwise the `arrow_shift` is not continous in `0..1` but instead at 1 the arrows jump back.

hdavid16 commented on src/beziercurves.jl:
```diff
@@ -67,6 +68,56 @@ end
+    let a = p0 - pt, b = p - p0
+    t = -(a[1]*b[1] + a[2]*b[2]) / (b[1]^2 + b[2]^2)
+"""
+function inverse_interpolate(p::BezierPath{PT}, pt) where PT
```
done

hdavid16 commented:
> I am not totally sure about the choice `arrow_shift=:end` vs `arrow_shift=1`. I tend to prefere `:end` because otherwise the `arrow_shift` is not continous in `0..1` but instead at 1 the arrows jump back.

I changed it to :end and updated docstring

hexaeder reviewed (commented):
Hey! I finally tried it locally and it works quite nice, thanks for the contribution! Three minor things i've noticed:

**Recalculate on `_size` change**
```julia
g = SimpleDiGraph(3)
add_edge!(g, 1, 1); add_edge!(g, 1, 2); add_edge!(g, 2, 1); add_edge!(g, 2, 3); add_edge!(g, 3, 1);

# test update of  node and arrow size
fig, ax, p = graphplot(g; arrow_shift=:end,
                       node_size=[20 for _ in 1:nv(g)],
                       arrow_size=[20 for _ in 1:ne(g)])

p.node_size[][1] = 40
notify(p.node_size)  # does not update arrow shift

p.arrow_size[][3] = 40
notify(p.arrow_size)  # does not update arrow shift

notify(p[:node_pos]) # updates arrow shift
```
I don't now the chain of observable trigger actions of the top of my head, but change in `[arrow|node]_size` should probably trigger update of the arrow shift.

**Error for 3d plos**
```julia
using GraphMakie: Spring
fig, ax, p = graphplot(g; arrow_shift=:end, layout=Spring(dim=3))
```
Errors at `reipes.jl:687` because of addition of 2d point to 3d point. I think this should fail early with an error message that 3d plots are not supportet with `:end` currently.

**interpolation fails when zooming out**
If the markers get very big with respect to the node distances something goes wrong and throws an error. The easisest to reproduce is to scroll out in a plot very far. I'm not sure where that happens. I think this is due to the desired distance getting negative or so. This should be chacked somewhere. Maybe let `reverse_interpolate` throw an `ArgumentError`, catch that in `update_arrowshift` and set it to 0.5?

hdavid16 commented on src/utils.jl:
```diff
@@ -11,7 +11,7 @@ end
     p = gp.plots[2]
     @assert p isa Scatter
-    @assert p.marker[] == '➤'
+    @assert p.marker[] == Arrow#'➤'
```
done

hdavid16 commented:
Thanks for testing these. 

- I fixed the lifted function so that it triggers with node size and arrow size changes.
- I now have 3D plots with `arrow_shift = :end` erroring at the `update_arrow_shift` function (I added a separate method for 3D plots). When we resolve #111, we will be able to update the code so that this arrow shift is supported for 3D plots.
- Interpolation now won't crash if we zoom out. The issue was occurring when the inverse interpolation gave a negative t value (see #112). I have now fixed this so that when a negative t value is passed, the first segment will be chosen for the interpolation (rather than the 0th or negative segment which doesn't exist). This is similar to what is done when t > 1. This fixed this error up to a certain point of zooming out. However, if you zoomed out far enough, the inverse interpolation (root finding with 5th order polynomial) fails because you start to get only imaginary roots or no roots at all). I now catch in `update_arrow_shift` when `t = NaN` and reset it to `0.5` with a `Warning`.
